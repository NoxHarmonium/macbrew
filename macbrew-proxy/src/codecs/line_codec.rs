use std::{io, str};
use tokio_util::codec::{Decoder, Encoder};

use bytes::BytesMut;

pub struct LineCodec;

impl Decoder for LineCodec {
    type Item = String;
    type Error = io::Error;

    fn decode(&mut self, src: &mut BytesMut) -> Result<Option<Self::Item>, Self::Error> {
        let newline = src.as_ref().iter().position(|b| *b == b'\n');
        if let Some(n) = newline {
            let line = src.split_to(n + 1);
            return match str::from_utf8(line.as_ref()) {
                Ok(s) => Ok(Some(s.to_string())),
                Err(_) => Err(io::Error::new(io::ErrorKind::Other, "Invalid String")),
            };
        }
        Ok(None)
    }
}

impl Encoder<String> for LineCodec {
    type Error = io::Error;

    fn encode(&mut self, item: String, dst: &mut BytesMut) -> Result<(), Self::Error> {
        Ok(())
    }
}
