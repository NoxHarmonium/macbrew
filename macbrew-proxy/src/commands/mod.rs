pub mod command;
pub mod get_fermentation;
pub mod get_recipes;
pub mod get_sessions;
pub mod list_sessions;
pub mod list_steps;
pub mod ping;
