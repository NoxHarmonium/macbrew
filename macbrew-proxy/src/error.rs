use reqwest::Error as ReqwestError;
use serde::{ser, Serialize};
use snafu::Snafu;

// TODO: Clean these up to provide more context and better error messages
#[derive(Debug, Snafu, Serialize)]
pub enum Error {
    #[snafu(display("Could not serialize data"))]
    SerializationFailure {},

    #[snafu(display("Error calling API: {}", inner_error))]
    ApiError {
        #[serde(skip_serializing)]
        inner_error: ReqwestError,
    },

    #[snafu(display("API response failed validation: {}", message))]
    ApiResponseValidationError { message: String },

    #[snafu(display("Invalid command input: {}", message))]
    InvalidCommandInput { message: String },

    #[snafu(display("Serial IO error: {}", inner_error))]
    SerialIoError {
        #[serde(skip_serializing)]
        inner_error: std::io::Error,
    },

    #[snafu(display("Unknown error: {}", message))]
    UnknownError { message: String },

    #[snafu(display("Could not allocate serialization buffer. (Not enough RAM?)"))]
    SerializeBufferFull {},

    #[snafu(display("Could not work out length of sequence"))]
    SerializeSeqLengthUnknown {},
}

pub type Result<T, E = Error> = std::result::Result<T, E>;

impl From<ReqwestError> for Error {
    fn from(inner_error: ReqwestError) -> Self {
        Error::ApiError {
            inner_error: inner_error,
        }
    }
}

impl From<std::io::Error> for Error {
    fn from(inner_error: std::io::Error) -> Self {
        Error::SerialIoError {
            inner_error: inner_error,
        }
    }
}

impl ser::Error for Error {
    #[cold]
    fn custom<T: std::fmt::Display>(msg: T) -> Error {
        // TODO: What should this actually be doing?
        // Hint: https://github.com/serde-rs/json/blob/master/src/error.rs#L373
        Error::UnknownError {
            message: msg.to_string(),
        }
    }
}
