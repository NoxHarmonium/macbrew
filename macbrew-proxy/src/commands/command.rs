use crate::data::data_manager::DataManager;
use crate::error::Result;
use crate::serializers::mac::serialize;
use async_trait::async_trait;
use serde::Serialize;

#[async_trait]
pub trait Command<DataManagerType: DataManager> {
    async fn handle(rid: &str, args: &[&str]) -> Result<Vec<u8>>;
}

pub fn prepare_response<T>(rid: &str, success: bool, response_object: &T) -> Result<Vec<u8>>
where
    T: Serialize + ?Sized,
{
    let preable_bytes = serialize(&rid)?;
    let success_bytes = serialize(&success)?;
    let response_bytes = serialize(&response_object)?;
    Ok([preable_bytes, success_bytes, response_bytes].concat())
}
