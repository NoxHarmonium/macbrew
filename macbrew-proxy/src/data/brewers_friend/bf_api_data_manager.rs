use crate::data::brewers_friend::bf_data_manager::BFDataManager;
use crate::data::brewers_friend::recipes::BeerXml;
use crate::data::brewers_friend::sessions::{BFBrewSession, BFBrewSessionsResponse};
use crate::error::{Error, Result};
use async_trait::async_trait;
use reqwest;
use reqwest::header::HeaderMap;
use serde::Deserialize;

#[cfg(test)]
use mockito;

#[derive(Deserialize, Debug)]
struct Config {
    brewers_friend_api_key: String,
}

pub struct BFApiDataManager;

fn get_base_url() -> String {
    #[cfg(test)]
    return mockito::server_url();
    #[cfg(not(test))]
    return String::from("https://api.brewersfriend.com");
}

fn create_client() -> Result<reqwest::Client> {
    let mut headers = HeaderMap::new();

    let config = match envy::from_env::<Config>() {
        Ok(config) => config,
        Err(error) => panic!("Missing environment variables: {:#?}", error),
    };

    headers.insert("X-API-KEY", config.brewers_friend_api_key.parse().unwrap());

    let client = reqwest::Client::builder()
        .default_headers(headers)
        .build()?;

    Ok(client)
}

#[async_trait]
impl BFDataManager for BFApiDataManager {
    async fn sessions() -> Result<Vec<BFBrewSession>> {
        println!("Fetching sessions...");
        let client = create_client()?;

        let brew_sessions_url = format!("{}/v1/brewsessions", get_base_url());

        let res = client.get(&brew_sessions_url).send().await?;

        let body = res.text().await?;

        let response: BFBrewSessionsResponse = match serde_json::from_str(&body) {
            Ok(response) => response,
            Err(error) => panic!(
                "Could not decode response from Brewers Friend API: {:#?}",
                error
            ),
        };

        println!("Fetched [{:#?}] sessions", response.brewsessions.len());

        Ok(response.brewsessions)
    }

    async fn session(id: &str) -> Result<BFBrewSession> {
        println!("Fetching session with id [{}]...", id);
        let client = create_client()?;

        let brew_session_url = format!("{}/v1/brewsessions/{}", get_base_url(), id);

        let res = client.get(&brew_session_url).send().await?;

        let body = res.text().await?;

        let response: BFBrewSessionsResponse = match serde_json::from_str(&body) {
            Ok(response) => response,
            Err(error) => panic!(
                "Could not decode response from Brewers Friend API: {:#?}",
                error
            ),
        };

        let session = match response.brewsessions.as_slice() {
            [] => Err(Error::ApiResponseValidationError {
                message: String::from("No session with ID found"),
            }),
            [session] => Ok(session.to_owned()),
            [_args @ ..] => Err(Error::ApiResponseValidationError {
                message: String::from("Multiple sessions for ID found. This should not happen"),
            }),
        };
        println!("Fetched session with id [{}]...", id);

        session
    }

    async fn recipe(recipe_id: &str) -> Result<BeerXml> {
        println!("Fetching recipe with id [{}]...", recipe_id);
        let client = create_client()?;

        let brew_session_url = format!("{}/v1/recipes/{}.xml", get_base_url(), recipe_id);

        let res = client.get(&brew_session_url).send().await?;

        let body = res.text().await?;

        let response = match quick_xml::de::from_str(&body) {
            Ok(response) => response,
            Err(error) => panic!(
                "Could not decode response from Brewers Friend API: {:#?}",
                error
            ),
        };

        println!("Fetched recipe with id [{}]...", recipe_id);

        Ok(response)
    }
}
