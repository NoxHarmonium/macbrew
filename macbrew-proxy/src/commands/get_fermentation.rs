use crate::commands::command::{prepare_response, Command};
use crate::data::brewers_friend::bf_data_manager::BfDataManager;
use crate::data::macbrew::fermentation::FermentationData;
use crate::error::Error::InvalidCommandInput;
use crate::error::Result;
use async_trait::async_trait;
use std::marker::PhantomData;

// TODO: Work out why rustc thinks T is unused. Is PhantomData needed? Can I restructure this?
// Clue: https://github.com/rust-lang/rust/issues/23246
pub struct GetFermentationDataCommand<DataManager: BfDataManager>(PhantomData<DataManager>);

#[async_trait]
impl<DataManager: BfDataManager + 'static> Command<DataManager>
    for GetFermentationDataCommand<DataManager>
{
    async fn handle(rid: &str, args: &[&str]) -> Result<Vec<u8>> {
        match args {
            [session_id] => {
                let bf_fermentation = DataManager::fermentation(session_id).await?;
                let fermentation_data = FermentationData::from_bf_fermentation(&bf_fermentation);
                prepare_response(rid, true, &fermentation_data)
            }
            [_args @ ..] => Err(InvalidCommandInput {
                message: String::from("Expected single parameter (Session ID)"),
            }),
        }
    }
}
