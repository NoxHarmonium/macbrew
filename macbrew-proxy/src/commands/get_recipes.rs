use crate::commands::command::{prepare_response, Command};
use crate::data::brewers_friend::bf_data_manager::BfDataManager;
use crate::data::macbrew::recipes::Recipe;
use crate::error::Error::InvalidCommandInput;
use crate::error::Result;
use async_trait::async_trait;
use std::marker::PhantomData;

// TODO: Work out why rustc thinks T is unused. Is PhantomData needed? Can I restructure this?
// Clue: https://github.com/rust-lang/rust/issues/23246
pub struct GetRecipesCommand<DataManager: BfDataManager>(PhantomData<DataManager>);

#[async_trait]
impl<DataManager: BfDataManager + 'static> Command<DataManager> for GetRecipesCommand<DataManager> {
    async fn handle(rid: &str, args: &[&str]) -> Result<Vec<u8>> {
        match args {
            [recipe_id] => {
                let beer_xml_recipe = DataManager::recipe(recipe_id).await?;
                let recipe = Recipe::from_beer_xml(&beer_xml_recipe);
                prepare_response(rid, true, &recipe)
            }
            [_args @ ..] => Err(InvalidCommandInput {
                message: String::from("Expected single parameter (Recipe ID)"),
            }),
        }
    }
}
