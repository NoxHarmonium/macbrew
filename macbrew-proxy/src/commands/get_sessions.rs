use crate::commands::command::{prepare_response, Command};
use crate::data::brewers_friend::bf_data_manager::BFDataManager;
use crate::data::macbrew::sessions::BrewSession;
use crate::error::Error::InvalidCommandInput;
use crate::error::Result;
use async_trait::async_trait;
use std::marker::PhantomData;

// TODO: Work out why rustc thinks T is unused. Is PhantomData needed? Can I restructure this?
// Clue: https://github.com/rust-lang/rust/issues/23246
pub struct GetSessionsCommand<DataManager: BFDataManager>(PhantomData<DataManager>);

#[async_trait]
impl<DataManager: BFDataManager + 'static> Command<DataManager>
    for GetSessionsCommand<DataManager>
{
    async fn handle(rid: &str, args: &[&str]) -> Result<Vec<u8>> {
        match args {
            [session_id] => {
                let bf_session = DataManager::session(session_id).await?;
                let session = BrewSession::from_bf_session(&bf_session);
                prepare_response(rid, true, &session)
            }
            [_args @ ..] => Err(InvalidCommandInput {
                message: String::from("Expected single parameter (Session ID)"),
            }),
        }
    }
}
