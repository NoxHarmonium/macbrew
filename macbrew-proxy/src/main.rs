#![warn(clippy::all)]

use clap::Clap;
use error::{Error, Result};
use futures::sink::SinkExt;
use std::str;
use tokio::stream::StreamExt;
use tokio_serial::{Serial, SerialPortSettings};
use tokio_util::codec::Framed;

mod codecs;
mod commands;
mod data;
mod error;
mod serializers;

#[cfg(unix)]
const DEFAULT_TTY: &str = "/dev/pts/1";
#[cfg(windows)]
const DEFAULT_TTY: &str = "COM1";

use codecs::line_codec::LineCodec;
use commands::command::{prepare_response, Command};
use data::brewers_friend::bf_api_data_manager::BFApiDataManager;

#[derive(Clap)]
#[clap(version = "0.1", author = "Sean Dawson <contact@seandawson.info>")]
struct Opts {
    #[clap(default_value = DEFAULT_TTY)]
    serial_device: String,
}

async fn handle_line(line: &str) -> Result<Vec<u8>> {
    let components = line.split(' ').collect::<Vec<_>>();
    let response = match components.as_slice() {
        [] => Err(Error::InvalidCommandInput {
            message: String::from(line),
        }),
        [rid, "LIST", "SESSION", args @ ..] => {
            commands::list_sessions::ListSessionsCommand::<BFApiDataManager>::handle(rid, args)
                .await
        }
        [rid, "GET", "SESSION", args @ ..] => {
            commands::get_sessions::GetSessionsCommand::<BFApiDataManager>::handle(rid, args).await
        }
        [rid, "GET", "RECIPE", args @ ..] => {
            commands::get_recipes::GetRecipesCommand::<BFApiDataManager>::handle(rid, args).await
        }
        [_, _args @ ..] => Err(Error::InvalidCommandInput {
            message: String::from(line),
        }),
    };
    match response {
        Ok(response) => Ok(response),
        Err(error) => prepare_response("ERROR", false, &error.to_string()),
    }
}

#[tokio::main]
async fn main() -> Result<()> {
    let opts: Opts = Opts::parse();

    println!("Starting MacBrew on port {}", opts.serial_device);

    let settings = SerialPortSettings::default();
    let mut port = Serial::from_path(opts.serial_device, &settings).unwrap();

    #[cfg(unix)]
    port.set_exclusive(false)
        .expect("Unable to set serial port exclusive to false");

    let mut framed = Framed::new(port, LineCodec);

    while let Some(line_result) = framed.next().await {
        let line = line_result.expect("Failed to read line");
        let response = handle_line(line.trim()).await?;
        framed.send(response).await?;
    }

    Ok(())
}

#[cfg(test)]
mod tests {

    use encoding::all::MAC_ROMAN;
    use encoding::{EncoderTrap, Encoding};
    use hexplay::HexViewBuilder;
    use mockito::mock;

    use super::*;

    fn assert_request_id(binary: &Vec<u8>, request_id: &'static str) -> () {
        assert_eq!(
            &binary[4..6],
            MAC_ROMAN
                .encode(request_id, EncoderTrap::Replace)
                .unwrap()
                .as_slice()
        );
    }

    fn format_binary_for_snap(binary: &Vec<u8>) -> Vec<String> {
        let view = HexViewBuilder::new(binary.as_slice())
            .row_width(16)
            .finish();
        // Split on new lines to make the snapshots easier to read
        view.to_string().split('\n').map(String::from).collect()
    }

    #[tokio::test]
    async fn test_get_recipe() {
        let recipe_xml = include_str!("../resources/sample_recipe.xml");
        let _m = mock("GET", "/v1/recipes/123456.xml")
            .with_status(200)
            .with_header("content-type", "application/xml")
            .with_header("x-api-key", "1234")
            .with_body(recipe_xml)
            .create();

        let result = handle_line("11 GET RECIPE 123456").await.unwrap();

        insta::assert_debug_snapshot!(format_binary_for_snap(&result));
        assert_request_id(&result, "11");
    }

    #[tokio::test]
    async fn test_list_sessions() {
        let sessions_json = include_str!("../resources/sample_sessions.json");
        let _m = mock("GET", "/v1/brewsessions")
            .with_status(200)
            .with_header("content-type", "application/xml")
            .with_header("x-api-key", "1234")
            .with_body(sessions_json)
            .create();

        let result = handle_line("22 LIST SESSION").await.unwrap();

        insta::assert_debug_snapshot!(format_binary_for_snap(&result));
        assert_request_id(&result, "22");
    }

    #[tokio::test]
    async fn test_get_sessions() {
        let session_json = include_str!("../resources/sample_session.json");
        let _m = mock("GET", "/v1/brewsessions/363597")
            .with_status(200)
            .with_header("content-type", "application/xml")
            .with_header("x-api-key", "1234")
            .with_body(session_json)
            .create();

        let result = handle_line("33 GET SESSION 363597").await.unwrap();

        insta::assert_debug_snapshot!(format_binary_for_snap(&result));
        assert_request_id(&result, "33");
    }
}
