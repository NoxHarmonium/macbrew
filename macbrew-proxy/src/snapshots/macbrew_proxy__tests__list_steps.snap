---
source: macbrew-proxy/src/main.rs
expression: format_binary_for_snap(&result)
---
[
    "00000000  05 F0 00 03 38 38 38 01 00 19 00 37 43 68 65 63  | ♣≡.♥888☺.↓.7Chec |",
    "00000010  6B 20 69 6E 67 72 65 64 69 65 6E 74 73 20 28 68  | k ingredients (h |",
    "00000020  6F 70 73 2F 79 65 61 73 74 20 65 74 63 2E 29 20  | ops/yeast etc.)  |",
    "00000030  61 6E 64 20 72 65 73 65 74 20 65 71 75 69 70 6D  | and reset equipm |",
    "00000040  65 6E 74 FF FF 10 00 24 50 72 65 70 61 72 65 20  | ent..►.$Prepare  |",
    "00000050  73 74 72 69 6B 65 20 77 61 74 65 72 20 28 61 64  | strike water (ad |",
    "00000060  64 20 73 61 6C 74 2F 61 63 69 64 29 FF FF 10 00  | d salt/acid)..►. |",
    "00000070  60 41 64 64 20 31 30 67 20 6F 66 20 22 41 6D 61  | `Add 10g of \"Ama |",
    "00000080  72 69 6C 6C 6F 20 28 38 2E 36 20 41 41 29 5C 72  | rillo (8.6 AA)\\r |",
    "00000090  5C 6E 20 20 20 20 20 20 20 20 20 20 20 20 20 20  | \\n               |",
    "000000A0  20 20 20 20 20 20 20 20 20 20 2F 20 31 34 30 20  |           / 140  |",
    "000000B0  47 72 61 6D 73 22 20 68 6F 70 73 20 74 6F 20 6D  | Grams\" hops to m |",
    "000000C0  61 73 68 20 61 74 20 6D 69 6E 75 74 65 20 22 30  | ash at minute \"0 |",
    "000000D0  22 00 00 12 00 66 41 64 64 20 35 67 20 6F 66 20  | \"..↕.fAdd 5g of  |",
    "000000E0  22 48 50 41 20 2D 20 20 45 63 6C 69 70 73 65 20  | \"HPA -  Eclipse  |",
    "000000F0  28 31 36 2E 39 20 41 41 29 5C 72 5C 6E 20 20 20  | (16.9 AA)\\r\\n    |",
    "00000100  20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20  |                  |",
    "00000110  20 20 20 20 20 2F 20 31 30 30 20 47 72 61 6D 73  |      / 100 Grams |",
    "00000120  22 20 68 6F 70 73 20 74 6F 20 6D 61 73 68 20 61  | \" hops to mash a |",
    "00000130  74 20 6D 69 6E 75 74 65 20 22 30 22 00 00 12 00  | t minute \"0\"..↕. |",
    "00000140  5F 41 64 64 20 32 30 67 20 6F 66 20 22 4C 65 6D  | _Add 20g of \"Lem |",
    "00000150  6F 6E 64 72 6F 70 20 28 36 20 41 41 29 5C 72 5C  | ondrop (6 AA)\\r\\ |",
    "00000160  6E 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20  | n                |",
    "00000170  20 20 20 20 20 20 20 20 20 2F 20 31 30 30 20 47  |          / 100 G |",
    "00000180  72 61 6D 73 22 20 68 6F 70 73 20 74 6F 20 6D 61  | rams\" hops to ma |",
    "00000190  73 68 20 61 74 20 6D 69 6E 75 74 65 20 22 30 22  | sh at minute \"0\" |",
    "000001A0  00 00 12 00 68 41 64 64 20 35 67 20 6F 66 20 22  | ..↕.hAdd 5g of \" |",
    "000001B0  4D 61 6E 64 61 72 69 6E 61 20 42 61 76 61 72 69  | Mandarina Bavari |",
    "000001C0  61 20 28 38 2E 35 20 41 41 29 5C 72 5C 6E 20 20  | a (8.5 AA)\\r\\n   |",
    "000001D0  20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20  |                  |",
    "000001E0  20 20 20 20 20 20 2F 20 31 30 30 20 47 72 61 6D  |       / 100 Gram |",
    "000001F0  73 22 20 68 6F 70 73 20 74 6F 20 6D 61 73 68 20  | s\" hops to mash  |",
    "00000200  61 74 20 6D 69 6E 75 74 65 20 22 30 22 00 00 12  | at minute \"0\"..↕ |",
    "00000210  00 23 48 65 61 74 20 22 53 74 72 69 6B 65 22 20  | .#Heat \"Strike\"  |",
    "00000220  77 61 74 65 72 20 74 6F 20 22 36 37 22 20 64 65  | water to \"67\" de |",
    "00000230  67 72 65 65 73 00 00 12 00 25 48 65 61 74 20 22  | grees..↕.%Heat \" |",
    "00000240  49 6E 66 75 73 69 6F 6E 22 20 77 61 74 65 72 20  | Infusion\" water  |",
    "00000250  74 6F 20 22 36 37 22 20 64 65 67 72 65 65 73 00  | to \"67\" degrees. |",
    "00000260  3C 12 00 25 48 65 61 74 20 22 49 6E 66 75 73 69  | <↕.%Heat \"Infusi |",
    "00000270  6F 6E 22 20 77 61 74 65 72 20 74 6F 20 22 37 35  | on\" water to \"75 |",
    "00000280  22 20 64 65 67 72 65 65 73 00 0A 12 00 25 48 65  | \" degrees.◙↕.%He |",
    "00000290  61 74 20 22 49 6E 66 75 73 69 6F 6E 22 20 77 61  | at \"Infusion\" wa |",
    "000002A0  74 65 72 20 74 6F 20 22 37 35 22 20 64 65 67 72  | ter to \"75\" degr |",
    "000002B0  65 65 73 00 00 12 00 2E 54 61 6B 65 20 67 72 61  | ees..↕..Take gra |",
    "000002C0  76 69 74 79 20 73 61 6D 70 6C 65 20 61 6E 64 20  | vity sample and  |",
    "000002D0  72 65 63 6F 72 64 20 70 72 65 2D 62 6F 69 6C 20  | record pre-boil  |",
    "000002E0  76 6F 6C 75 6D 65 FF FF 12 00 0F 42 72 69 6E 67  | volume..↕.☼Bring |",
    "000002F0  20 74 6F 20 61 20 62 6F 69 6C FF FF 12 00 2F 41  |  to a boil..↕./A |",
    "00000300  64 64 20 31 67 20 6F 66 20 22 57 61 72 72 69 6F  | dd 1g of \"Warrio |",
    "00000310  72 22 20 68 6F 70 73 20 74 6F 20 62 6F 69 6C 20  | r\" hops to boil  |",
    "00000320  61 74 20 6D 69 6E 75 74 65 20 22 31 30 22 00 0A  | at minute \"10\".◙ |",
    "00000330  14 00 0F 54 75 72 6E 20 6F 66 66 20 6B 65 74 74  | ¶.☼Turn off kett |",
    "00000340  6C 65 00 0A 14 00 13 53 74 61 72 74 20 63 68 69  | le.◙¶.‼Start chi |",
    "00000350  6C 6C 69 6E 67 20 77 6F 72 74 FF FF 14 00 62 57  | lling wort..¶.bW |",
    "00000360  68 65 6E 20 74 65 6D 70 20 69 73 20 53 6F 6D 65  | hen temp is Some |",
    "00000370  28 22 37 35 22 29 20 64 65 67 72 65 65 73 2C 20  | (\"75\") degrees,  |",
    "00000380  61 64 64 20 31 30 67 20 6F 66 20 22 41 6D 61 72  | add 10g of \"Amar |",
    "00000390  69 6C 6C 6F 20 28 38 2E 36 20 41 41 29 22 20 68  | illo (8.6 AA)\" h |",
    "000003A0  6F 70 73 20 74 6F 20 77 68 69 72 6C 70 6F 6F 6C  | ops to whirlpool |",
    "000003B0  20 66 6F 72 20 22 31 35 22 20 6D 69 6E 75 74 65  |  for \"15\" minute |",
    "000003C0  73 FF FF 16 00 90 57 68 65 6E 20 74 65 6D 70 20  | s..▬.ÉWhen temp  |",
    "000003D0  69 73 20 53 6F 6D 65 28 22 37 35 22 29 20 64 65  | is Some(\"75\") de |",
    "000003E0  67 72 65 65 73 2C 20 61 64 64 20 31 30 67 20 6F  | grees, add 10g o |",
    "000003F0  66 20 22 48 50 41 20 2D 20 20 45 63 6C 69 70 73  | f \"HPA -  Eclips |",
    "00000400  65 20 28 31 36 2E 39 20 41 41 29 5C 72 5C 6E 20  | e (16.9 AA)\\r\\n  |",
    "00000410  20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20  |                  |",
    "00000420  20 20 20 20 20 20 20 2F 20 31 30 30 20 47 72 61  |        / 100 Gra |",
    "00000430  6D 73 22 20 68 6F 70 73 20 74 6F 20 77 68 69 72  | ms\" hops to whir |",
    "00000440  6C 70 6F 6F 6C 20 66 6F 72 20 22 31 35 22 20 6D  | lpool for \"15\" m |",
    "00000450  69 6E 75 74 65 73 FF FF 16 00 61 57 68 65 6E 20  | inutes..▬.aWhen  |",
    "00000460  74 65 6D 70 20 69 73 20 53 6F 6D 65 28 22 37 35  | temp is Some(\"75 |",
    "00000470  22 29 20 64 65 67 72 65 65 73 2C 20 61 64 64 20  | \") degrees, add  |",
    "00000480  31 30 67 20 6F 66 20 22 4C 65 6D 6F 6E 64 72 6F  | 10g of \"Lemondro |",
    "00000490  70 20 28 36 20 41 41 29 22 20 68 6F 70 73 20 74  | p (6 AA)\" hops t |",
    "000004A0  6F 20 77 68 69 72 6C 70 6F 6F 6C 20 66 6F 72 20  | o whirlpool for  |",
    "000004B0  22 31 35 22 20 6D 69 6E 75 74 65 73 FF FF 16 00  | \"15\" minutes..▬. |",
    "000004C0  92 57 68 65 6E 20 74 65 6D 70 20 69 73 20 53 6F  | ÆWhen temp is So |",
    "000004D0  6D 65 28 22 37 35 22 29 20 64 65 67 72 65 65 73  | me(\"75\") degrees |",
    "000004E0  2C 20 61 64 64 20 31 30 67 20 6F 66 20 22 4D 61  | , add 10g of \"Ma |",
    "000004F0  6E 64 61 72 69 6E 61 20 42 61 76 61 72 69 61 20  | ndarina Bavaria  |",
    "00000500  28 38 2E 35 20 41 41 29 5C 72 5C 6E 20 20 20 20  | (8.5 AA)\\r\\n     |",
    "00000510  20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20  |                  |",
    "00000520  20 20 20 20 2F 20 31 30 30 20 47 72 61 6D 73 22  |     / 100 Grams\" |",
    "00000530  20 68 6F 70 73 20 74 6F 20 77 68 69 72 6C 70 6F  |  hops to whirlpo |",
    "00000540  6F 6C 20 66 6F 72 20 22 31 35 22 20 6D 69 6E 75  | ol for \"15\" minu |",
    "00000550  74 65 73 FF FF 16 00 1C 43 6F 6F 6C 20 77 6F 72  | tes..▬.∟Cool wor |",
    "00000560  74 20 64 6F 77 6E 20 74 6F 20 53 6F 6D 65 28 22  | t down to Some(\" |",
    "00000570  33 37 22 29 FF FF 16 00 14 54 61 6B 65 20 67 72  | 37\")..▬.¶Take gr |",
    "00000580  61 76 69 74 79 20 72 65 61 64 69 6E 67 FF FF 16  | avity reading..▬ |",
    "00000590  00 15 54 72 61 6E 73 66 65 72 20 74 6F 20 66 65  | .§Transfer to fe |",
    "000005A0  72 6D 65 6E 74 6F 72 FF FF 16 00 0B 50 69 74 63  | rmentor..▬.♂Pitc |",
    "000005B0  68 20 79 65 61 73 74 FF FF 16 00 1D 53 65 74 20  | h yeast..▬.↔Set  |",
    "000005C0  66 65 72 6D 65 6E 74 61 74 69 6F 6E 20 63 68 61  | fermentation cha |",
    "000005D0  6D 62 65 72 20 74 65 6D 70 FF FF 18 00 11 43 6C  | mber temp..↑.◄Cl |",
    "000005E0  65 61 6E 75 70 20 65 71 75 69 70 6D 65 6E 74 FF  | eanup equipment. |",
    "000005F0  FF 18 96 81 9B EF                                | .↑ûüø∩           |",
]
